src/lib/address_resolve/tool.cpp

```diff
@@ -0,0 +1,164 @@
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AddressResolve.h"
+
+#include <lib/support/CHIPMem.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/CHIPDeviceLayer.h>
+
+#include <condition_variable>
+#include <mutex>
+#include <stdlib.h>
+#include <strings.h>
+
+namespace {
+
+using namespace chip;
+using chip::Inet::InterfaceId;
+using chip::Transport::PeerAddress;
+
+// clang-format off
+const char * const sHelp =
+    "Usage: address-resolve-tool <command> [ <args...> ]\n"
+    "\n"
+    "Commands:\n"
+    "\n"
+    "    node <nodeid> <fabricid> -- Find the node for the given node/fabric.\n"
+    "\n";
+// clang-format on
+
+class PrintOutNodeListener : public chip::AddressResolve::NodeListener
+{
+public:
+    PrintOutNodeListener() { mSelfHandle.SetListener(this); }
+
+    void OnNodeAddressResolved(const PeerId & peerId, const PeerAddress & address) override
+    {
+        char addr_string[PeerAddress::kMaxToStringSize];
+        address.ToString(addr_string);
+
+        ChipLogProgress(Discovery, "Resolve completed: %s", addr_string);
+        NotifyDone();
+    }
+
+    void OnNodeAddressResolutionFailed(const PeerId & peerId, CHIP_ERROR reason) override
+    {
+        ChipLogError(Discovery, "Failed to resolve: %s", reason.AsString());
+        NotifyDone();
+    }
+
+    AddressResolve::NodeLookupHandle & Handle() { return mSelfHandle; }
+
+    void NotifyDone() { DeviceLayer::PlatformMgr().StopEventLoopTask(); }
+
+private:
+    AddressResolve::NodeLookupHandle mSelfHandle;
+    bool mDone = false;
+};
+
+PrintOutNodeListener gListener;
+
+CHIP_ERROR Initialize()
+{
+    ReturnErrorOnFailure(DeviceLayer::PlatformMgr().InitChipStack());
+    ReturnErrorOnFailure(Dnssd::Resolver::Instance().Init(DeviceLayer::UDPEndPointManager()));
+    ReturnErrorOnFailure(AddressResolve::Resolver::Instance().Init(&DeviceLayer::SystemLayer()));
+    return CHIP_NO_ERROR;
+}
+
+void Shutdown()
+{
+    Dnssd::Resolver::Instance().Shutdown();
+    DeviceLayer::PlatformMgr().Shutdown();
+}
+
+bool Cmd_Node(int argc, const char ** argv)
+{
+    if (argc != 2)
+    {
+        ChipLogError(NotSpecified, "Please provide nodeid and fabricid as arguments");
+        return false;
+    }
+
+    uint64_t node = strtoull(argv[0], nullptr, 0 /* auto-detect base */);
+
+    if (node == 0)
+    {
+        ChipLogError(NotSpecified, "Node ID %s not supported", argv[0]);
+        return false;
+    }
+
+    uint64_t fabric = strtoull(argv[1], nullptr, 0 /* auto-detect base */);
+    if (node == 0)
+    {
+        ChipLogError(NotSpecified, "Fabric ID %s not supported", argv[1]);
```

Comment:
This is not a fabric Id. It's a compressed fabric ID. Let's not call it fabricID anywhere, since operational discovery relies on comrpessed fabric ID. Since we do not use a fabric table here, the callers must have computed the exact compressed fabric ID, and pass it it. Please rename any mention of `fabric id` to `compressed fabric reference`